src/pystow/impl.py

```diff
@@ -178,6 +178,43 @@ def ensure(
         )
         return path
 
+    def ensure_custom(
+        self,
+        *subkeys: str,
+        name: str,
+        force: bool = False,
+        provider: Callable[..., None],
+        **kwargs,
+    ) -> Path:
+        """Ensure a file is present, and run a custom create function otherwise.
+
+        :param subkeys:
+            A sequence of additional strings to join. If none are given,
+            returns the directory for this module.
+        :param name:
+            The file name.
+        :param force:
+            Should the file be re-created, even if the path already exists?
+        :param provider:
+            The file provider. Will be run with the path as the first positional argument,
+            if the file needs to be generated.
+        :param kwargs:
+            Additional keyword-based parameters passed to the provider.
+
+        :raises ValueError:
+            If the provider was called but the file was not created by it.
+
+        :return:
+            The path of the file that has been created (or already exists)
+        """
+        path = self.join(*subkeys, name=name, ensure_exists=True)
+        if path.is_file() and not force:
+            return path
+        provider(path, **kwargs)
+        if not path.is_file():
+            raise ValueError(f"Provider {provider} did not create the file at {path}!")
```

Comment:
why not insist the provider function returns some object (e.g., a boolean or class) that contains information about how the download went?

Reply:
this would require to wrap custom functions to adhere to the common API. The result would still be the same, right? If the download / creation failed, we cannot recover from that.

Reply:
yeah. so it's up to you